synapse/storage/databases/state/bg_updates.py

```diff
@@ -98,81 +98,115 @@ def _get_state_groups_from_groups_txn(
+                else:
+                    for state_key in state_keys:
+                        state_filter_condition_combos.append((etype, state_key))
+        # And here is the optimization itself. We don't want to do the optimization
+        # if there are too many individual conditions. 10 is an arbitrary number
+        # with no testing behind it but we do know that we specifically made this
+        # optimization for when we grab the necessary state out for
+        # `filter_events_for_client` which just uses 2 conditions
+        # (`EventTypes.RoomHistoryVisibility` and `EventTypes.Member`).
+        if use_condition_optimization and len(state_filter_condition_combos) < 10:
+            select_clause_list: List[str] = []
+            for etype, skey in state_filter_condition_combos:
+                if skey is None:
+                    where_clause = "(type = ?)"
+                    overall_select_query_args.extend([etype])
+                else:
+                    where_clause = "(type = ? AND state_key = ?)"
+                    overall_select_query_args.extend([etype, skey])
+
+                # Small helper function to wrap the union clause in parenthesis if we're
+                # using postgres. This is because SQLite doesn't allow `LIMIT`/`ORDER`
+                # clauses in the union subquery but postgres does as long as they are
+                # wrapped in parenthesis which this function handles the complexity of
+                # handling.
+                def wrap_union_if_postgres(
+                    union_clause: str, extra_order_or_limit_clause: str = ""
+                ) -> str:
+                    if isinstance(self.database_engine, PostgresEngine):
+                        return f"""({union_clause} {extra_order_or_limit_clause})"""
+
+                    return union_clause
+
+                # We could use `SELECT DISTINCT ON` here to align with the query below
+                # but that isn't compatible with SQLite and we can get away with `LIMIT
+                # 1` here instead because the `WHERE` clause will only ever match and
+                # target one event; and is simpler anyway. And it's better to use
+                # something that's simpler and compatible with both Database engines.
+                select_clause_list.append(
+                    wrap_union_if_postgres(
+                        # We only select `state_group` here for use in the `ORDER`
+                        # clause later after the `UNION`
                         f"""
-                        (
-                            SELECT DISTINCT ON (type, state_key)
-                                type, state_key, event_id
-                            FROM state_groups_state
-                            INNER JOIN sgs USING (state_group)
-                            WHERE {where_clause}
-                            ORDER BY type, state_key, state_group DESC
-                        )
```

Comment:
One improvement that I thought I could take out of this was simplifying these conditions like the following but it seems to perform about the same (at least in my one example I use against `matrix.org`) -> https://explain.depesz.com/s/pWH9

```suggestion
                        (
                            SELECT type, state_key, event_id
                            FROM state_groups_state
                            INNER JOIN sgs USING (state_group)
                            WHERE {where_clause}
                            ORDER BY type, state_key, state_group DESC
                            LIMIT 1
                        )
```

<details>
<summary>Raw query and <code>EXPLAIN ANALYZE</code></summary

```sql
EXPLAIN ANALYZE WITH RECURSIVE sgs(state_group) AS (
    VALUES(739988088::bigint)
    UNION ALL
    SELECT prev_state_group FROM state_group_edges e, sgs s
    WHERE s.state_group = e.state_group
)
(
    SELECT type, state_key, event_id, state_group
    FROM state_groups_state
    INNER JOIN sgs USING (state_group)
    WHERE (type = 'm.room.history_visibility' AND state_key = '')
    ORDER BY type, state_key, state_group DESC
    LIMIT 1
)
 UNION 
(
    SELECT type, state_key, event_id, state_group
    FROM state_groups_state
    INNER JOIN sgs USING (state_group)
    WHERE (type = 'm.room.member' AND state_key = '@madlittlemods:matrix.org')
    ORDER BY type, state_key, state_group DESC
    LIMIT 1
);
                                                                                                QUERY PLAN
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Unique  (cost=1272.13..1272.15 rows=2 width=104) (actual time=1.707..1.717 rows=2 loops=1)
   CTE sgs
     ->  Recursive Union  (cost=0.00..284.28 rows=101 width=8) (actual time=0.004..0.504 rows=59 loops=1)
           ->  Result  (cost=0.00..0.01 rows=1 width=8) (actual time=0.002..0.002 rows=1 loops=1)
           ->  Nested Loop  (cost=0.57..28.23 rows=10 width=8) (actual time=0.007..0.007 rows=1 loops=59)
                 ->  WorkTable Scan on sgs s  (cost=0.00..0.20 rows=10 width=8) (actual time=0.000..0.000 rows=1 loops=59)
                 ->  Index Only Scan using state_group_edges_unique_idx on state_group_edges e  (cost=0.57..2.79 rows=1 width=16) (actual time=0.006..0.006 rows=1 loops=59)
                       Index Cond: (state_group = s.state_group)
                       Heap Fetches: 0
   ->  Sort  (cost=987.85..987.85 rows=2 width=104) (actual time=1.705..1.708 rows=2 loops=1)
         Sort Key: state_groups_state.type, state_groups_state.state_key, state_groups_state.event_id, state_groups_state.state_group
         Sort Method: quicksort  Memory: 25kB
         ->  Append  (cost=493.90..987.84 rows=2 width=104) (actual time=1.136..1.545 rows=2 loops=1)
               ->  Limit  (cost=493.90..493.90 rows=1 width=91) (actual time=1.134..1.136 rows=1 loops=1)
                     ->  Sort  (cost=493.90..493.90 rows=1 width=91) (actual time=1.133..1.134 rows=1 loops=1)
                           Sort Key: state_groups_state.state_group DESC
                           Sort Method: quicksort  Memory: 25kB
                           ->  Nested Loop  (cost=0.84..493.89 rows=1 width=91) (actual time=1.111..1.122 rows=1 loops=1)
                                 ->  CTE Scan on sgs  (cost=0.00..2.02 rows=101 width=8) (actual time=0.007..0.532 rows=59 loops=1)
                                 ->  Index Scan using state_groups_state_type_idx on state_groups_state  (cost=0.84..4.86 rows=1 width=91) (actual time=0.009..0.009 rows=0 loops=59)
                                       Index Cond: ((state_group = sgs.state_group) AND (type = 'm.room.history_visibility'::text) AND (state_key = ''::text))
               ->  Limit  (cost=493.90..493.90 rows=1 width=91) (actual time=0.405..0.406 rows=1 loops=1)
                     ->  Sort  (cost=493.90..493.90 rows=1 width=91) (actual time=0.404..0.405 rows=1 loops=1)
                           Sort Key: state_groups_state_1.state_group DESC
                           Sort Method: quicksort  Memory: 25kB
                           ->  Nested Loop  (cost=0.84..493.89 rows=1 width=91) (actual time=0.398..0.399 rows=1 loops=1)
                                 ->  CTE Scan on sgs sgs_1  (cost=0.00..2.02 rows=101 width=8) (actual time=0.000..0.013 rows=59 loops=1)
                                 ->  Index Scan using state_groups_state_type_idx on state_groups_state state_groups_state_1  (cost=0.84..4.86 rows=1 width=91) (actual time=0.006..0.006 rows=0 loops=59)
                                       Index Cond: ((state_group = sgs_1.state_group) AND (type = 'm.room.member'::text) AND (state_key = '@madlittlemods:matrix.org'::text))
 Planning Time: 7.594 ms
 Execution Time: 3.023 ms
(31 rows)
```

</details>


Compare to the *after* query performance from the previous PR, https://github.com/matrix-org/synapse/pull/14527 -> https://explain.depesz.com/s/Qi4A

Reply:
I guess I already explored this in https://github.com/matrix-org/synapse/pull/14494#discussion_r1029636023 and came to the same conclusion